src/keyszer/config_api.py

```diff
@@ -41,6 +41,27 @@
 # multipurpose timeout
 _TIMEOUTS = TIMEOUT_DEFAULTS
 
+# global dict of delay values used to mitigate Unicode entry sequence and macro or combo failures
+THROTTLE_DELAY_DEFAULTS = {
+    'key_pre_delay_ms': 0,
+    'key_post_delay_ms': 0,
+}
+_THROTTLES = THROTTLE_DELAY_DEFAULTS
+
+def throttle_delays(key_pre_delay_ms=0, key_post_delay_ms=0):
+    _kpre, _kpost, ms_min, ms_max = key_pre_delay_ms, key_post_delay_ms, 0.0, 150.0
```

Comment:
Please lose the `_kpre, _kpost,` variables, there is nothing wrong with writing out the longer ones.  Also surely Python have a `min` and `max` function we could use to simplify this?

You can write a `withinRange` function easily enough if a language gives you min and max... then just call `clamp_to_range(0,150,x)`, etc...  then these are just simple assignments.  Do we truly need the errors?

Reply:
I don't understand the point of writing a new function just to check whether a variable is within a range, rather than using the simple comparison. 

Errors? Are you objecting to alerting the user they are giving the API a value out of range? How else are they going to know why their out-of-range value doesn't seem to be working? 

Do you mean you want to just force it to be zero if the given value is below zero (which is what it defaults to now), and also force it to be the max value if the value is above the max? And never show an error?

Reply:
It dries up the complexity here... much simpler to read.

```py
VALID_RANGE = range(0,150)
error(..) if delay not in VALID_RANGE
_THROTTLE[...] = clamp_to(delay, VALID_RANGE)
```

Reply:
Ha, Python 3.10 already includes `clamp`:

```
result = clamp(7, 2, 10)
print(result)  # Output: 7
```

Reply:
My expert advisor is telling me that `range()` returns a "range object" not a list of integers. Although apparently in some circumstances this operates similarly to a Python list of integers, I don't think you can just directly use the `not in` syntax with it. At the very least it would need to be converted to a list with `list()`. 

And in Python, a one-line `if` (aka, the Python version of a ternary operator) always needs an `else`. I made certain that was the case, because I wanted to do something similar to that second line a couple of times. 

Am I understanding correctly.  that one of the main things you want out of these changes is for the API to "partially" succeed if the user input is out-of-range? Instead of just always defaulting to zero?

Reply:
`clamp()`

Ah, that's handy. I didn't see your other comment until I refreshed.

Reply:
Do you have PyTorch on your system? I don't think Python in general has a built-in `clamp()` function yet. Everything I see online says you need to define it yourself, although it is pretty simple and lots of pages online show it as an example. 

I have Python 3.10 on my system(s) and there is no such function when I try to use it in VSCode. 🤔

Reply:
I dunno, OpenGPT told me - I only have 3.9.6 local on my laptop so I can't test.

> At the very least it would need to be converted to a list with list().

Nope, this is a common use of ranges.

```
>>> 3 in range(1,5)
True
```

It only works for integers though.

Reply:
Is this more like what you want? 

```py
def clamp(num, min_value, max_value):
    return max(min(num, max_value), min_value)

def throttle_delays(key_pre_delay_ms=0, key_post_delay_ms=0):
    ms_min, ms_max = 0.0, 150.0
    if any(key_pre_delay_ms, key_post_delay_ms):
        error(f'Throttle delay value out of range. Clamping to closest value between {ms_min} and {ms_max}.')
    _THROTTLES['key_pre_delay_ms'] = clamp(key_pre_delay_ms, ms_min, ms_max)
    _THROTTLES['key_post_delay_ms'] = clamp(key_post_delay_ms, ms_min, ms_max)
```

Reply:
> It only works for integers though.

Makes sense. But I'm allowing floats so the user can go between zero and 1ms.

Reply:
I didn't realize that `update()` is mainly for if you want to update multiple dict values at the same time. I could do both keys on one line that way, but it would be a pretty long line. I'd rather not.

Reply:
Yes, but I don't know how `  if any(key_pre_delay_ms, key_post_delay_ms):` works...

Update could be nice, _remember a single statement can be spread about multiple lines._. The keymaps are often a single statement... but organized on many lines.

Reply:
I believe ChatGPT is drawing on some bad information. Even importing the "math" module, there is no built-in `clamp()` function. I found a couple of Python issues where people were asking for a `clamp()` function to be added as a built-in, and suggestions of adding it to "math", but those issues were closed with no apparent agreement on adding the function, even to "math". 

I think you still have to define it yourself, like I did above.

Reply:
I do spread statements across multiple lines when necessary. I just like to avoid it when possible. 

Fixed the `any()` line in the example. Forgot it wants a list, and needed to be checking for out-of-range, not just not-zero. Works now.

Reply:
>  I just like to avoid it when possible.

Why?

Reply:
Really need to print out the values of the throttle delays from `config_api`, since `output.py` is evaluated too early and there is still the issue of the values being zero. 

Here is a version using `update()` and passing in a dict instead of assigning on separate lines. I don't really like it as much as the earlier example. 

```py
def clamp(num, min_value, max_value):
    return max(min(num, max_value), min_value)

def throttle_delays(key_pre_delay_ms=0, key_post_delay_ms=0):
    ms_min, ms_max = 0.0, 150.0
    if ms_min > any([key_pre_delay_ms, key_post_delay_ms]) > ms_max:
        error(f'Throttle delay value out of range. Clamping to closest value between {ms_min} and {ms_max}.')
    _THROTTLES.update({
        'key_pre_delay_ms': clamp(key_pre_delay_ms, ms_min, ms_max), 
        'key_post_delay_ms': clamp(key_post_delay_ms, ms_min, ms_max),
    })
    if any(_THROTTLES.values()) > 0:
        debug(f'THROTTLES: {key_pre_delay_ms = }, {key_post_delay_ms = }')
```

Reply:
Well first, lets get it working.  :-) I don't think you understand what `any` does, maybe try playing around with it in a console...  might help reading a little bit about parsing or precedence any how things are evaluated...

`any(_THROTTLES.values()) ` is an expression, so it's evaluated FIRST... then you have either:

`False > 0` or `True > 0` which isn't what you are trying to say at all.  If it works here it's because 0 is false and 1 is truthy it has nothing to do with the `> 0 ` at all...

Same for your earlier use... any is it's own expression... it has nothing to do with your greater or lesser than comparisons. To do these kinds of things in python you may want to look into list comprehensions.

Reply:
Python surely has helper functions like Ruby that would take a lambda: https://apidock.com/ruby/Enumerable/any%3F

Reply:
Ah. I see what you're saying. I've used it correctly in other situations like `matchProps()`, probably because I was following examples more closely. 

This should actually do what I was intending. 

```py
if any([elem > 0 for elem in _THROTTLES.values()]):
```

It's already been clamped, so when this runs it will show the throttles if any are not zero.

Reply:
Yeah, that looks very Pythonic. :-) `any` simple function paired with the list comp.

Reply:
Testing seems to indicate that I finally found most of the things I screwed up on every line in the previous examples: 

```py
def throttle_delays(key_pre_delay_ms=0, key_post_delay_ms=0):
    ms_min, ms_max = 0.0, 150.0
    if any([not(ms_min <= e <= ms_max) for e in [key_pre_delay_ms, key_post_delay_ms]]):
        error(f'Throttle delay value out of range. Clamping to closest value between {ms_min} and {ms_max}.')
    _THROTTLES.update({
        'key_pre_delay_ms': clamp(key_pre_delay_ms, ms_min, ms_max), 
        'key_post_delay_ms': clamp(key_post_delay_ms, ms_min, ms_max),
    })
    if any([elem > 0 for elem in _THROTTLES.values()]) > 0:
        debug(f'THROTTLES:\
            \n\tPre-key  : {_THROTTLES["key_pre_delay_ms"]}\
            \n\tPost-key : {_THROTTLES["key_post_delay_ms"]}')
```

Reply:
Yeah, looks much better.  Why do we need the last conditional though?  If someone is calling `throttle_delays` at all I think we can assume it's with non-zero values...

Reply:
>  If someone is calling throttle_delays at all I think we can assume it's with non-zero values

Hm. Did not think of it that way. This will only run if the user calls it. Right. 

Actually, to be more precise, the conditional will block the debug output if the user chooses to set the values to zero, which is possible/allowed. I was thinking about that earlier, but figured that would be a rare event. 

Even better, without the conditional, it will display the values if the user calls the function at all, even if they are set to zero. By the user. So it removes a contradiction. 

Good catch. 

```py
def throttle_delays(key_pre_delay_ms=0, key_post_delay_ms=0):
    ms_min, ms_max = 0.0, 150.0
    if any([not(ms_min <= e <= ms_max) for e in [key_pre_delay_ms, key_post_delay_ms]]):
        error(f'Throttle delay value out of range. Clamping to valid range: {ms_min} to {ms_max}.')
    _THROTTLES.update({ 'key_pre_delay_ms': clamp(key_pre_delay_ms, ms_min, ms_max), 
                        'key_post_delay_ms': clamp(key_post_delay_ms, ms_min, ms_max) })
    debug(f'THROTTLES:\
        \n\tPre-key  : {_THROTTLES["key_pre_delay_ms"]}\
        \n\tPost-key : {_THROTTLES["key_post_delay_ms"]}')
```